### DISCUSSION TITLE
Improve type-checking inside the engine source code

### DISCUSSION THREAD

**Discussion by mvaligursky** [playcanvas/engine#3358] [General]
In the engine source code, we only import modules that are required in order to build the engine. This leaves many types unknown to Visual Studio Code (or similar JS editor), and we see the type with red underscore, and when typing `this.device.` we don't get any code-completion aids. This is what we see:

<img width="691" alt="Screenshot 2021-07-22 at 14 23 23" src="https://user-images.githubusercontent.com/59932779/126646436-8c418bbe-685c-4b65-8237-ce724db13169.png">

We could adopt the use of typescript typedef imports to improve the situation. First, we'd need to import the type this way:

<img width="741" alt="Screenshot 2021-07-22 at 14 27 15" src="https://user-images.githubusercontent.com/59932779/126646622-100566d5-9804-476b-924c-aca322a0f23b.png">

Then declare _this.device_ to be of type _GraphicsDevice_

And then Visual Studio gives us code-completion and understands and validates the function signatures.

<img width="678" alt="Screenshot 2021-07-22 at 14 28 32" src="https://user-images.githubusercontent.com/59932779/126646990-facd7097-075a-4af4-9ce6-b709ea093adc.png">

(Thanks to @vkalpias who's started to use it in Editor API he's working on)

**Comment by yaustar** [comment]
Does this mean we add a bunch of typedefs to a lot of the files? Does this affect JS docs in any way? Should we include all types that are needed anyway? Not just the minimum to build the engine?

As a side note, Webstorm shows this as:
![image](https://user-images.githubusercontent.com/16639049/126648001-a26c7957-368e-42ee-b27e-9185b0c88471.png)

**Reply by mvaligursky** [reply]
I don't believe it has any effect apart from improving our coding experience. These imports are only for parsing of code for intellisense, nothing else.

We cannot import all types, as that would create circular dependencies. You only need to import types you need to know a type of in the code (because you call constructor, or call static function on the type, or do typeof .. and similar). Otherwise imports are marked as not needed anyways.

**Comment by mvaligursky** [comment]
It would be also great to find a way we can have all typefeds in a single file for the project, that gets included using a single line into any source files. Perhaps something like this:
https://github.com/polyforest/jsdoc-tsimport-plugin

**Comment by kungfooman** [comment]
> We cannot import all types, as that would create circular dependencies.

TypeScript is analyzing code dependencies and removes dispensable (type) imports, so technically you probably just need to put `typescript()` before the JavaScript code is processed by rollup (at least I imagine that would work).

I did a little test for just this one file example of yours:

![image](https://user-images.githubusercontent.com/5236548/126656579-90803e76-c863-4029-8679-3b13e5515833.png)

I added this line: 

```js
import { GraphicsDevice } from 'src/graphics/graphics-device.js';
```

And type definitions where TS/VSCode expects them (differs from strange JSDoc style over the class itself):

```js
class ForwardRenderer {
    /**
     * @type {GraphicsDevice}
     */
    device;

    /**
     * 
     * @param {GraphicsDevice} graphicsDevice 
     */
    constructor(graphicsDevice) {
        this.device = graphicsDevice;
        var device = this.device;
```

![image](https://user-images.githubusercontent.com/5236548/126657110-3a2dc286-41ca-4c7f-8fa8-aad165836cfa.png)

And then just compile it for comparison: `npx tsc --allowJs --outDir tsctest -m ESNext src/scene/renderer/forward-renderer.js`
